hal/ia32/memory.c

```diff
@@ -66,51 +79,141 @@ int hal_memoryGetEntry(unsigned int *offs, mem_entry_t *entry)
 		"1: "
 		"addl $0xc, %%esp; "
 	: "+a" (ret), "+b" (next)
-	: "D" (entry)
+	: "D" (&mem)
 	: "ecx", "edx", "memory", "cc");
+
 	*offs = next;
 
+	entry->start = mem.addr;
+	/* TODO: mapent_t should use addr & len instead of start & end */
+	entry->end = (mem.addr + mem.len <= 0xffffffff) ? mem.len + mem.addr : mem.addr + mem.len - 1;
+	entry->type = hal_entryReserved;
+
+	*biosType = mem.type;
+
 	return ret;
 }
 
 
-int hal_memAddMap(addr_t start, addr_t end, u32 attr, u32 mapId)
+static void hal_getMinOverlappedRange(addr_t start, addr_t end, mapent_t *entry, mapent_t *minEntry)
 {
-	return EOK;
+	if ((start < entry->end) && (end > entry->start)) {
+		if (start > entry->start)
+			entry->start = start;
+
+		if (end < entry->end)
+			entry->end = end;
+
+		if (entry->start < minEntry->start) {
+			minEntry->start = entry->start;
+			minEntry->end = entry->end;
+			minEntry->type = entry->type;
+		}
+	}
 }
 
 
-int hal_memoryInit(void)
+int hal_memoryGetNextEntry(addr_t start, addr_t end, mapent_t *entry)
 {
-	memory_enableA20();
+	unsigned int i, offs, biosType;
+	mapent_t tempEntry = { 0 }, prevEntry = { 0 };
+	mapent_t minEntry = { .start = (addr_t)-1, .end = 0 };
+
+	static const mapent_t entries[] = {
+		{ .start = 0x0, .end = 0x1000, .type = hal_entryTemp },
+		{ .start = (addr_t)_plo_bss, .end = (addr_t)_end, .type = hal_entryTemp },
+		{ .start = ADDR_STACK, .end = ADDR_STACK + SIZE_STACK, .type = hal_entryTemp },
+		{ .start = ADDR_STACK, .end = ADDR_STACK + SIZE_STACK, .type = hal_entryTemp },
+		{ .start = ADDR_GDT, .end = ADDR_GDT + SIZE_GDT, .type = hal_entryTemp },
+		{ .start = ADDR_IDT, .end = ADDR_IDT + SIZE_IDT, .type = hal_entryTemp },
```

Comment:
Why do we mark these entries (except plo bss) as temporary (`hal_entryTemp`)? Shouldn't they be marked as reserved?

Reply:
The appropriate comment has been added to the code. 
They are used only by plo it is why `.type =  hal_entryTemp`.  Kernel gets information about these specific entries from `hal_syspage_t`.